### PR TITLE
issue 301 coin leaderboard resolved

### PR DESCRIPTION
**Pull request by probro27**
## Summary of Changes
Changed the model of `UserCoinEntry` to add a rank field and updated the SQL query to get the rank of the user. Updated the `getUserIdCurrentCoinPosition` to get the current rank of the user based on the Rank() function of SQL. 

## Related Issues
To resolve the issue #301 

## Demonstration of Changes
image

### CODE REVIEW

**Review by mcpenguin — CHANGES_REQUESTED** [review]
Hi, thanks for doing this! I'm not too familiar with how OVER works in SQLite, but it seems that there's a bug?
image
During local testing, even though I'm ranked 3rd, the "your position" still says I'm first.

**Comment by probro27** [comment]
Oh I'm sorry. Yeah I only tested with 2 users, one me and one test account. That's why I could have missed this. Let me fix it.

**Comment by probro27** [comment]
Should be fine now. 

image

**Review by mcpenguin — CHANGES_REQUESTED** [review]
Hmm, I still see a bug: if we there is an unknown user in the leaderboard (ie if you do <@[id]> and it says "Unknown User"), it is not displayed in the leaderboard, but the rankings still account for them.
image
For example, there is an "unknown user" in 5th place, so the user in 6th place should really be 5th, instead of 6th.

**Comment by probro27** [comment]
Oh didn't know "Unknown user" was a thing. I got it, will work on it.

File: src/commandDetails/coin/leaderboard.ts [hunk 1]
```diff
@@ -47,9 +48,16 @@ const getCurrentCoinLeaderboardEmbed = async (
...
-    const userCoinEntryText = `${leaderboardArray.length + 1}. ${cleanUserTag} - ${
+    const userCoinEntryText = `${rank}. ${cleanUserTag} - ${
       userCoinEntry.balance
     } ${getCoinEmoji()}`;
+    if (userCoinEntry.user_id === currentUserId) {
+      currentPosition = rank;
+    }
+    if (previousBalance !== userCoinEntry.balance) {
+      previousBalance = userCoinEntry.balance;
+      rank = rank + 1;
```
Comment (Picowchew) [inline src/commandDetails/coin/leaderboard.ts:59]: By this logic, the leaderboard would look something like the following:
```
1. User1 - 400
2. User2 - 300
3. User3 - 300
3. User4 - 300
3. User5 - 200
4. User6 - 100
```

File: src/components/coin.ts [hunk 2]
```diff
@@ -67,6 +67,7 @@ export const coinBonusMap = new Map<BonusType, Bonus>([
 export interface UserCoinEntry {
   user_id: string;
   balance: number;
+  rank: number;
```
Comment (Picowchew) [inline src/components/coin.ts:70]: Since rank is now manually computed, this doesn't seem to be needed?

File: src/components/coin.ts [hunk 3]
```diff
@@ -150,32 +151,17 @@ export const getCurrentCoinLeaderboard = async (limit = 10): Promise<UserCoinEnt
   const db = await openDB();
   const res = await db.all(
     `
-      SELECT user_id, balance
+      SELECT user_id, balance,
+      Rank() OVER (ORDER BY balance DESC) AS rank
       FROM user_coin
-      ORDER BY balance DESC
+      ORDER BY rank ASC
```
Comment (Picowchew) [inline src/components/coin.ts:157]: Once again, since rank is now manually computed, the previous simpler logic seems to be fine?

**Review by Picowchew — CHANGES_REQUESTED** [review]

File: src/components/coin.ts (hunk 3 above)
Comment (probro27) [reply src/components/coin.ts]: Fixed it in this commit

File: src/components/coin.ts (hunk 2 above)
Comment (probro27) [reply src/components/coin.ts]: Fixed it in this commit.

File: src/commandDetails/coin/leaderboard.ts (hunk 1 above)
Comment (probro27) [reply src/commandDetails/coin/leaderboard.ts:56]: Updated the logic to display this correctly

**Comment by probro27** [comment]
Hope this works @Picowchew

File: src/commandDetails/coin/leaderboard.ts [hunk 4]
```diff
@@ -47,9 +48,16 @@ const getCurrentCoinLeaderboardEmbed = async (
       .join('\\_')
       .split('`')
       .join('\\`');
-    const userCoinEntryText = `${leaderboardArray.length + 1}. ${cleanUserTag} - ${
+    const userCoinEntryText = `${rank}. ${cleanUserTag} - ${
```
Comment (Picowchew) [inline src/commandDetails/coin/leaderboard.ts:51]: The leaderboard would currently show the person in 1st place to have a rank of 0?

**Review by Picowchew — CHANGES_REQUESTED** [review]
The implementation still looks buggy. It'd be good to run your bot and test your code?

File: src/commandDetails/coin/leaderboard.ts (hunk 4 above)
Comment (probro27) [reply src/commandDetails/coin/leaderboard.ts:51]: So sorry. I didn't change much last time so I thought I got it correct. Just updated it and checked it.
image
The ranking should be good now. @Picowchew

**Comment by probro27** [comment]
@Picowchew So sorry. I just seem to forget about it. Let me just fix it and update the PR.

**Review by Picowchew — APPROVED** [review]
LGTM!

**Review by mcpenguin — APPROVED** [review]
LGTM too!

**Pull request merged**